### PR TITLE
Error handling in handleRedirectCallback

### PR DESCRIPTION
**Pull request by davidpatrick**
Throw errors provided by Auth0 authentication callback query params

Resolves https://github.com/auth0/auth0-spa-js/issues/24

### CODE REVIEW

File: src/global.ts [hunk 1]
```diff
@@ -177,12 +177,22 @@ interface LogoutOptions {
   client_id?: string;
 }
 
+/**
+ * @ignore
+ */
+interface AuthenticationErrorOptions {
```
Comment (luisrudge) [inline src/global.ts]: what's this for?

File: src/errors.ts [hunk 2]
```diff
@@ -0,0 +1,6 @@
+export class AuthenticationError extends Error {
+  constructor(name: string, message: string) {
+    super(message);
+    this.name = name;
```
Comment (luisrudge) [inline src/errors.ts]: I think I rather have both `error` and `error_description` in the error.

```
super(error_description);
this.error = error;
this.error_description = error_description;
```

File: __tests__/index.test.ts [hunk 3]
```diff
@@ -386,6 +388,27 @@ describe('Auth0', () => {
...
+      it('throws error with AuthenticationError', async () => {
+        const { auth0, utils } = await localSetup();
+        const queryResult = { error: 'unauthorized' };
+        utils.parseQueryResult.mockReturnValue(queryResult);
+
+        await expect(auth0.handleRedirectCallback()).rejects.toBeInstanceOf(
+          AuthenticationError
+        );
+      });
+      it('throws error with message from error_description', async () => {
```
Comment (luisrudge) [inline __tests__/index.test.ts]: there's no test to validate we're adding the `error` property to the AuthenticationError object

File: src/errors.ts (hunk 2 above)
Comment (davidpatrick) [reply src/errors.ts]: Do you recommend not extending the Error class?  I've mapped `error_description` to `message` and `error` to the name.

Comment (luisrudge) [reply src/errors.ts]: We should absolutely extend the error class, but we should keep the original properties (`error` and `error_description`) intact so the dev can do:

```
catch ({error, error_description}) {
  //handle it
}
```

Comment (luisrudge) [reply src/errors.ts]: Also, if there's more information in the error that we receive (like `state` or something else), they should be here as well.

File: src/errors.ts [hunk 4]
```diff
@@ -0,0 +1,12 @@
+export class AuthenticationError extends Error {
+  error: string;
+  error_description: string;
+  state: string;
+
+  constructor(error: string, error_description?: string, state?: string) {
```
Comment (luisrudge) [inline src/errors.ts]: 😬typescript tip of the day 😬 

you can add `pulic` or `private` in the constructor args to automatically create the fields in the class:

```js
class CustomError extends Error {
  constructor(
    public error: string,
    public error_description: string,
    public state: string
  ) {
    super(error_description);
  }
}

```

Comment (davidpatrick) [reply src/errors.ts]: Thanks @luisrudge !

**Pull request merged**